From 5a4e8c5752a9d3c9e481c16efa794cc8ab5425c4 Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Wed, 19 Mar 2025 19:04:30 +0100
Subject: [PATCH] Issue 6626 - Ignore replica busy condition in healthcheck
 (#6630)

Replica Busy condition is expected when there is more than 2 suppliers so healthcheck should not report any error for such condition.
Fixed issue in CI tests:

test_healthcheck_replication_out_of_sync_not_broken was unstable and redundant with test_healthcheck_replica_busy so I moved it back in health_repl_test.py and rewrite it to test working replication whose replica are not in sync (healthcheck should not report anything)
some tests (not always the same) were randomly failing with an empty log (which is unexpected because healthcheck output should never be empty.
I suspect the log capture mechanism so health_repl_test.py now run dsctl instance healthcheck using subprocess module and capture the output. So far I have not changed the other files because I have not noticed any failure.
Issue: #6626

Reviewed by: @tbordaz (Thanks!)

(cherry picked from commit bc22cfa6184f51b8492c692f1c95e721d538ab5e)
---
 .../suites/healthcheck/health_repl_test.py    | 245 ++++++++++++++----
 .../suites/healthcheck/health_sync_test.py    | 130 ----------
 src/lib389/lib389/replica.py                  |   3 +
 3 files changed, 191 insertions(+), 187 deletions(-)
 delete mode 100644 dirsrvtests/tests/suites/healthcheck/health_sync_test.py

diff --git a/dirsrvtests/tests/suites/healthcheck/health_repl_test.py b/dirsrvtests/tests/suites/healthcheck/health_repl_test.py
index a8d94dfcb..729cc3d6b 100644
--- a/dirsrvtests/tests/suites/healthcheck/health_repl_test.py
+++ b/dirsrvtests/tests/suites/healthcheck/health_repl_test.py
@@ -9,56 +9,124 @@
 
 import pytest
 import os
-from contextlib import suppress
+import random
+import re
+import string
+import subprocess
+import threading
+import time
+from contextlib import suppress, AbstractContextManager
 from lib389.backend import Backend, Backends
 from lib389.idm.user import UserAccounts
 from lib389.replica import Changelog, ReplicationManager, Replicas
 from lib389.utils import *
 from lib389._constants import *
 from lib389.cli_base import FakeArgs
-from lib389.topologies import topology_m2, topology_m3
 from lib389.cli_ctl.health import health_check_run
+from lib389.topologies import topology_m2, topology_m3
 from lib389.paths import Paths
 
 CMD_OUTPUT = 'No issues found.'
 JSON_OUTPUT = '[]'
 
+LOGIC_DICT = {
+        False: ( "not ", "", lambda x: x ),
+        True:  ( "", "not ", lambda x: not x )
+    }
+
 ds_paths = Paths()
 log = logging.getLogger(__name__)
 
 
-def run_healthcheck_and_flush_log(topology, instance, searched_code, json, searched_code2=None):
-    args = FakeArgs()
-    args.instance = instance.serverid
-    args.verbose = instance.verbose
-    args.list_errors = False
-    args.list_checks = False
-    args.check = ['replication', 'backends:userroot:cl_trimming']
-    args.dry_run = False
-
+class LoadInstance(AbstractContextManager):
+    @staticmethod
+    def create_test_user(inst):
+        users = UserAccounts(inst, DEFAULT_SUFFIX)
+        uid = str(20000 + int(inst.serverid[8:]))
+        properties = {
+            'uid': f'testuser_{inst.serverid}',
+            'cn' : f'testuser_{inst.serverid}',
+            'sn' : 'user_{inst.serverid}',
+            'uidNumber' : uid,
+            'gidNumber' : uid,
+            'homeDirectory' : f'/home/testuser_{inst.serverid}'
+        }
+        return users.ensure_state(properties=properties)
+
+    def __init__(self, inst):
+        self.inst = inst
+        self.stop = threading.Event()
+        self.thread = threading.Thread(target=self.loader)
+        self.user = LoadInstance.create_test_user(inst)
+
+    def loader(self):
+        while not self.stop.is_set():
+            value = ''.join(random.choices(string.ascii_uppercase + string.digits, k=10))
+            self.user.replace('description', value)
+            #log.info(f'Modified {self.user.dn} description with {value} on {self.inst.serverid}')
+            time.sleep(0.001)
+
+    def __exit__(self, *args):
+        self.stop.set()
+        self.thread.join()
+        self.user.delete()
+
+    def __enter__(self):
+        self.thread.start()
+        return self
+
+
+class BreakReplication(AbstractContextManager):
+    def __init__(self, inst):
+        self.replica = Replicas(inst).list()[0]
+        self.oldval = None
+
+    def __exit__(self, *args):
+        self.replica.replace('nsds5ReplicaBindDNGroup', self.oldval)
+
+    def __enter__(self):
+        self.oldval = self.replica.get_attr_val_utf8('nsds5ReplicaBindDNGroup')
+        self.replica.replace('nsds5ReplicaBindDNGroup', 'cn=repl')
+        return self
+
+
+def assert_is_in_result(result, searched_code, isnot=False):
+    # Assert if searched_code is not in logcap
+    if searched_code is None:
+        return
+    
+    # Handle positive and negative tests:
+    nomatch, match, f = LOGIC_DICT[bool(isnot)]
+    try:
+        assert f(re.search(re.escape(searched_code), result))
+        log.info(f'Searched code {searched_code} is {match}in healthcheck output')
+    except AssertionError as exc:
+        log.error(f'{searched_code} is {nomatch}in healthcheck output:  {result}')
+        raise
+
+
+def run_healthcheck_and_check_result(topology, instance, searched_code, json, searched_code2=None, isnot=False):
+    cmd = [ 'dsctl', ]
     if json:
-        log.info('Use healthcheck with --json option')
-        args.json = json
-        health_check_run(instance, topology.logcap.log, args)
-        assert topology.logcap.contains(searched_code)
-        log.info('Healthcheck returned searched code: %s' % searched_code)
-
-        if searched_code2 is not None:
-            assert topology.logcap.contains(searched_code2)
-            log.info('Healthcheck returned searched code: %s' % searched_code2)
-    else:
-        log.info('Use healthcheck without --json option')
-        args.json = json
-        health_check_run(instance, topology.logcap.log, args)
-        assert topology.logcap.contains(searched_code)
-        log.info('Healthcheck returned searched code: %s' % searched_code)
-
-        if searched_code2 is not None:
-            assert topology.logcap.contains(searched_code2)
-            log.info('Healthcheck returned searched code: %s' % searched_code2)
-
-    log.info('Clear the log')
-    topology.logcap.flush()
+        cmd.append('--json')
+        if searched_code == CMD_OUTPUT:
+            searched_code = JSON_OUTPUT
+    cmd.append(instance.serverid)
+    cmd.extend(['healthcheck', '--check', 'replication' , 'backends:userroot:cl_trimming'])
+
+    result = subprocess.run(cmd, capture_output=True, universal_newlines=True)
+    log.info(f'Running: {cmd}')
+    log.info(f'Stdout: {result.stdout}')
+    log.info(f'Stderr: {result.stdout}')
+    log.info(f'Return code: {result.returncode}')
+    stdout = result.stdout
+
+    # stdout should not be empty
+    assert stdout is not None
+    assert len(stdout) > 0
+    assert_is_in_result(stdout, searched_code, isnot=isnot)
+    assert_is_in_result(stdout, searched_code2, isnot=isnot)
+
 
 
 def set_changelog_trimming(instance):
@@ -112,15 +180,15 @@ def test_healthcheck_replication_replica_not_reachable(topology_m2):
     with suppress(Exception):
         repl.wait_for_replication(M1, M2, timeout=5)
 
-    run_healthcheck_and_flush_log(topology_m2, M1, RET_CODE, json=False)
-    run_healthcheck_and_flush_log(topology_m2, M1, RET_CODE, json=True)
+    run_healthcheck_and_check_result(topology_m2, M1, RET_CODE, json=False)
+    run_healthcheck_and_check_result(topology_m2, M1, RET_CODE, json=True)
 
     log.info('Set nsds5replicaport for the replication agreement to a reachable port')
     agmt_m1.replace('nsDS5ReplicaPort', '{}'.format(M2.port))
     repl.wait_for_replication(M1, M2)
 
-    run_healthcheck_and_flush_log(topology_m2, M1, CMD_OUTPUT, json=False)
-    run_healthcheck_and_flush_log(topology_m2, M1, JSON_OUTPUT, json=True)
+    run_healthcheck_and_check_result(topology_m2, M1, CMD_OUTPUT, json=False)
+    run_healthcheck_and_check_result(topology_m2, M1, JSON_OUTPUT, json=True)
 
 
 @pytest.mark.xfail(ds_is_older("1.4.1"), reason="Not implemented")
@@ -160,13 +228,13 @@ def test_healthcheck_changelog_trimming_not_configured(topology_m2):
 
     time.sleep(3)
 
-    run_healthcheck_and_flush_log(topology_m2, M1, RET_CODE, json=False)
-    run_healthcheck_and_flush_log(topology_m2, M1, RET_CODE, json=True)
+    run_healthcheck_and_check_result(topology_m2, M1, RET_CODE, json=False)
+    run_healthcheck_and_check_result(topology_m2, M1, RET_CODE, json=True)
 
     set_changelog_trimming(M1)
 
-    run_healthcheck_and_flush_log(topology_m2, M1, CMD_OUTPUT, json=False)
-    run_healthcheck_and_flush_log(topology_m2, M1, JSON_OUTPUT, json=True)
+    run_healthcheck_and_check_result(topology_m2, M1, CMD_OUTPUT, json=False)
+    run_healthcheck_and_check_result(topology_m2, M1, JSON_OUTPUT, json=True)
 
 
 @pytest.mark.xfail(ds_is_older("1.4.1"), reason="Not implemented")
@@ -208,8 +276,8 @@ def test_healthcheck_replication_presence_of_conflict_entries(topology_m2):
 
     repl.test_replication_topology(topology_m2)
 
-    run_healthcheck_and_flush_log(topology_m2, M1, RET_CODE, json=False)
-    run_healthcheck_and_flush_log(topology_m2, M1, RET_CODE, json=True)
+    run_healthcheck_and_check_result(topology_m2, M1, RET_CODE, json=False)
+    run_healthcheck_and_check_result(topology_m2, M1, RET_CODE, json=True)
 
 
 def test_healthcheck_non_replicated_suffixes(topology_m2):
@@ -245,6 +313,44 @@ def test_healthcheck_non_replicated_suffixes(topology_m2):
     health_check_run(inst, topology_m2.logcap.log, args)
 
 
+def test_healthcheck_replica_busy(topology_m3):
+    """Check that HealthCheck does not returns DSREPLLE0003 code when a replicva is busy
+
+    :id: b7c4a5aa-ef98-11ef-87f5-482ae39447e5
+    :setup: 3 MMR topology
+    :steps:
+        1. Create a 3 suppliers full-mesh topology
+        2. Generate constant modify load on S1 and S2
+        3. Wait a bit to ensure stable replication flow
+        4. Perform a modify on S3
+        5. Use HealthCheck on S3 without --json option
+        6. Use HealthCheck on S3 with --json option
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Success
+        4. Success
+        5. Healthcheck should not reports DSREPLLE0003 code and related details
+        6. Healthcheck should not reports DSREPLLE0003 code and related details
+    """
+
+    RET_CODE = 'DSREPLLE0003'
+    # Is DSREPLLE0003 ignored if replica is busy ?
+    ignored = not ds_is_older("2.7")
+
+    S1 = topology_m3.ms['supplier1']
+    S2 = topology_m3.ms['supplier2']
+    S3 = topology_m3.ms['supplier3']
+    with LoadInstance(S1), LoadInstance(S2):
+        # Wait a bit to let replication starts
+        time.sleep(10)
+        # Create user on S3 then remove it:
+        LoadInstance(S3).user.delete()
+        # S3 agrements should now be in the replica busy state
+        run_healthcheck_and_check_result(topology_m3, S3, RET_CODE, json=False, isnot=ignored)
+        run_healthcheck_and_check_result(topology_m3, S3, RET_CODE, json=True, isnot=ignored)
+
+
 @pytest.mark.xfail(ds_is_older("1.4.1"), reason="Not implemented")
 def test_healthcheck_replication_out_of_sync_broken(topology_m3):
     """Check if HealthCheck returns DSREPLLE0001 code
@@ -265,25 +371,50 @@ def test_healthcheck_replication_out_of_sync_broken(topology_m3):
 
     RET_CODE = 'DSREPLLE0001'
 
-    M1 = topology_m3.ms['supplier1']
-    M2 = topology_m3.ms['supplier2']
-    M3 = topology_m3.ms['supplier3']
+    S1 = topology_m3.ms['supplier1']
+    S2 = topology_m3.ms['supplier2']
+    S3 = topology_m3.ms['supplier3']
 
     log.info('Break supplier2 and supplier3')
-    replicas = Replicas(M2)
-    replica = replicas.list()[0]
-    replica.replace('nsds5ReplicaBindDNGroup', 'cn=repl')
+    with BreakReplication(S2), BreakReplication(S3):
+        time.sleep(1)
+        log.info('Perform update on supplier1')
+        test_users_m1 = UserAccounts(S1, DEFAULT_SUFFIX)
+        test_users_m1.create_test_user(1005, 2000)
 
-    replicas = Replicas(M3)
-    replica = replicas.list()[0]
-    replica.replace('nsds5ReplicaBindDNGroup', 'cn=repl')
+        time.sleep(3)
+        run_healthcheck_and_check_result(topology_m3, S1, RET_CODE, json=False)
+        run_healthcheck_and_check_result(topology_m3, S1, RET_CODE, json=True)
 
-    log.info('Perform update on supplier1')
-    test_users_m1 = UserAccounts(M1, DEFAULT_SUFFIX)
-    test_users_m1.create_test_user(1005, 2000)
 
-    run_healthcheck_and_flush_log(topology_m3, M1, RET_CODE, json=False)
-    run_healthcheck_and_flush_log(topology_m3, M1, RET_CODE, json=True)
+def test_healthcheck_replication_out_of_sync_not_broken(topology_m3):
+    """Check that HealthCheck returns no issues when replication is in progress
+
+    :id: 8305000d-ba4d-4c00-8331-be0e8bd92150
+    :setup: 3 MMR topology
+    :steps:
+        1. Create a 3 suppliers full-mesh topology, all replicas being synchronized
+        2. Generate constant load on two supplier
+        3. Use HealthCheck without --json option
+        4. Use HealthCheck with --json option
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Healthcheck reports no issue found
+        4. Healthcheck reports no issue found
+    """
+
+    RET_CODE = CMD_OUTPUT
+
+    S1 = topology_m3.ms['supplier1']
+    S2 = topology_m3.ms['supplier2']
+    S3 = topology_m3.ms['supplier3']
+
+    with LoadInstance(S1), LoadInstance(S2):
+        # Wait a bit to let replication starts
+        time.sleep(10)
+        run_healthcheck_and_check_result(topology_m3, S1, RET_CODE, json=False)
+        run_healthcheck_and_check_result(topology_m3, S1, RET_CODE, json=True)
 
 
 if __name__ == '__main__':
diff --git a/dirsrvtests/tests/suites/healthcheck/health_sync_test.py b/dirsrvtests/tests/suites/healthcheck/health_sync_test.py
deleted file mode 100644
index 8f9bc0a00..000000000
--- a/dirsrvtests/tests/suites/healthcheck/health_sync_test.py
+++ /dev/null
@@ -1,130 +0,0 @@
-# --- BEGIN COPYRIGHT BLOCK ---
-# Copyright (C) 2020 Red Hat, Inc.
-# All rights reserved.
-#
-# License: GPL (version 3 or any later version).
-# See LICENSE for details.
-# --- END COPYRIGHT BLOCK ---
-#
-
-import pytest
-import os
-import time
-from datetime import *
-from lib389.idm.user import UserAccounts
-from lib389.utils import *
-from lib389._constants import *
-from lib389.cli_base import FakeArgs
-from lib389.topologies import topology_m3
-from lib389.cli_ctl.health import health_check_run
-from lib389.paths import Paths
-
-ds_paths = Paths()
-log = logging.getLogger(__name__)
-
-
-def run_healthcheck_and_flush_log(topology, instance, searched_code, json, searched_code2=None):
-    args = FakeArgs()
-    args.instance = instance.serverid
-    args.verbose = instance.verbose
-    args.list_errors = False
-    args.list_checks = False
-    args.check = ['replication']
-    args.dry_run = False
-
-    if json:
-        log.info('Use healthcheck with --json option')
-        args.json = json
-        health_check_run(instance, topology.logcap.log, args)
-        assert topology.logcap.contains(searched_code)
-        log.info('Healthcheck returned searched code: %s' % searched_code)
-
-        if searched_code2 is not None:
-            assert topology.logcap.contains(searched_code2)
-            log.info('Healthcheck returned searched code: %s' % searched_code2)
-    else:
-        log.info('Use healthcheck without --json option')
-        args.json = json
-        health_check_run(instance, topology.logcap.log, args)
-        assert topology.logcap.contains(searched_code)
-        log.info('Healthcheck returned searched code: %s' % searched_code)
-
-        if searched_code2 is not None:
-            assert topology.logcap.contains(searched_code2)
-            log.info('Healthcheck returned searched code: %s' % searched_code2)
-
-    log.info('Clear the log')
-    topology.logcap.flush()
-
-
-# This test is in separate file because it is timeout specific
-@pytest.mark.xfail(ds_is_older("1.4.1"), reason="Not implemented")
-#unstable or unstatus tests, skipped for now
-@pytest.mark.flaky(max_runs=2, min_passes=1)
-def test_healthcheck_replication_out_of_sync_not_broken(topology_m3):
-    """Check if HealthCheck returns DSREPLLE0003 code
-
-    :id: 8305000d-ba4d-4c00-8331-be0e8bd92150
-    :setup: 3 MMR topology
-    :steps:
-        1. Create a 3 suppliers full-mesh topology, all replicas being synchronized
-        2. Stop M1
-        3. Perform an update on M2 and M3.
-        4. Check M2 and M3 are synchronized.
-        5. From M2, reinitialize the M3 agreement
-        6. Stop M2 and M3
-        7. Restart M1
-        8. Start M3
-        9. Use HealthCheck without --json option
-        10. Use HealthCheck with --json option
-    :expectedresults:
-        1. Success
-        2. Success
-        3. Success
-        4. Success
-        5. Success
-        6. Success
-        7. Success
-        8. Success
-        9. Healthcheck reports DSREPLLE0003 code and related details
-        10. Healthcheck reports DSREPLLE0003 code and related details
-    """
-
-    RET_CODE = 'DSREPLLE0003'
-
-    M1 = topology_m3.ms['supplier1']
-    M2 = topology_m3.ms['supplier2']
-    M3 = topology_m3.ms['supplier3']
-
-    log.info('Stop supplier1')
-    M1.stop()
-
-    log.info('Perform update on supplier2 and supplier3')
-    test_users_m2 = UserAccounts(M2, DEFAULT_SUFFIX)
-    test_users_m3 = UserAccounts(M3, DEFAULT_SUFFIX)
-    test_users_m2.create_test_user(1000, 2000)
-    for user_num in range(1001, 3000):
-        test_users_m3.create_test_user(user_num, 2000)
-    time.sleep(2)
-
-    log.info('Stop M2 and M3')
-    M2.stop()
-    M3.stop()
-
-    log.info('Start M1 first, then M2, so that M2 acquires M1')
-    M1.start()
-    M2.start()
-    time.sleep(2)
-
-    log.info('Start M3 which should not be able to acquire M1 since M2 is updating it')
-    M3.start()
-    time.sleep(2)
-
-    run_healthcheck_and_flush_log(topology_m3, M3, RET_CODE, json=False)
-    run_healthcheck_and_flush_log(topology_m3, M3, RET_CODE, json=True)
-
-
-if __name__ == '__main__':
-    # Run isolated
-    # -s for DEBUG mode
-    CURRENT_FILE = os.path.realpath(__file__)
diff --git a/src/lib389/lib389/replica.py b/src/lib389/lib389/replica.py
index 0486c48f8..0dc3a0f54 100644
--- a/src/lib389/lib389/replica.py
+++ b/src/lib389/lib389/replica.py
@@ -1276,6 +1276,9 @@ class Replica(DSLdapObject):
                                 report['check'] = f'replication:agmts_status'
                                 yield report
                         elif status['state'] == 'amber':
+                            if "can't acquire busy replica" in status['reason']:
+                                # Ignore replica busy condition
+                                continue
                             # Warning
                             report = copy.deepcopy(DSREPLLE0003)
                             report['detail'] = report['detail'].replace('SUFFIX', suffix)
-- 
2.49.0

